From 9757d05e27ac2114179185aee5630f235d3968a5 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 26 Aug 2024 14:31:08 +0200
Subject: [PATCH 1099/1695] drivers: stm32_etzpc: restrain access on non secure
 peripherals for OP-TEE

As per previous behavior for the firewall bus probe, do not allow non
secure peripherals to be probed in OP-TEE except for some listed
exceptions. The firewall mechanism exception only applies to the secure
level as bad CID filtering configuration is prohibited by hardware.

To do so, implement a driver specific firewall bus probe that will
only probe secure peripherals and implement firewall exceptions
when CFG_INSECURE is set for which no firewall operations will be done.
This allows, for example, to share a console with the non-secure world
for development purposes.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9838a419132681d87aec343fdd360ac5b66dd467 drivers: stm32_etzpc: restrain access on non secure peripherals for OP-TEE]
Change-Id: Ie57ecb1d2b357f4b471ec683dc9f805c1c1c1661
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/400277
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/arch/arm/plat-stm32mp1/main.c       | 25 +++++++++
 core/arch/arm/plat-stm32mp1/stm32_util.h |  2 +
 core/drivers/firewall/stm32_etzpc.c      | 71 +++++++++++++++++++++++-
 3 files changed, 96 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 34a6844c0..0ef864e64 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -790,3 +790,28 @@ TEE_Result stm32_activate_internal_tamper(int id)
 
 	return res;
 };
+
+bool stm32mp_allow_probe_shared_device(const void *fdt, int node)
+{
+	static int uart_console_node = -1;
+	const char *compat = NULL;
+	static bool once;
+
+	if (!once) {
+		get_console_node_from_dt((void *)fdt, &uart_console_node,
+					 NULL, NULL);
+		once = true;
+	}
+
+	compat = fdt_stringlist_get(fdt, node, "compatible", 0, NULL);
+
+	/*
+	 * Allow OP-TEE console and MP15 I2C to be shared
+	 * with non-secure world
+	 */
+	if (node == uart_console_node ||
+	    !strcmp(compat, "st,stm32mp15-i2c-non-secure"))
+		return true;
+
+	return false;
+}
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index a21f37a04..6cedcf4cd 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -344,4 +344,6 @@ void __noreturn do_reset(const char *str);
 
 TEE_Result stm32_activate_internal_tamper(int id);
 
+bool stm32mp_allow_probe_shared_device(const void *fdt, int node);
+
 #endif /*__STM32_UTIL_H__*/
diff --git a/core/drivers/firewall/stm32_etzpc.c b/core/drivers/firewall/stm32_etzpc.c
index 0257475df..0a51eb258 100644
--- a/core/drivers/firewall/stm32_etzpc.c
+++ b/core/drivers/firewall/stm32_etzpc.c
@@ -395,7 +395,8 @@ static TEE_Result stm32_etzpc_acquire_access(struct firewall_query *firewall)
 	id = firewall->args[0] & ETZPC_ID_MASK;
 	if (id < etzpc_dev->ddata->num_per_sec) {
 		attr = etzpc_do_get_decprot(etzpc_dev, id);
-		if (attr == ETZPC_DECPROT_MCU_ISOLATION)
+		if (attr != ETZPC_DECPROT_S_RW &&
+		    attr != ETZPC_DECPROT_NS_R_S_W)
 			return TEE_ERROR_ACCESS_DENIED;
 	} else {
 		return TEE_ERROR_BAD_PARAMETERS;
@@ -606,6 +607,72 @@ static void fdt_etzpc_conf_decprot(struct etzpc_device *dev,
 	clk_disable(dev->pdata.clk);
 }
 
+static TEE_Result
+stm32_etzpc_dt_probe_bus(const void *fdt, int node,
+			 struct firewall_controller *ctrl __maybe_unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct firewall_query *fw = NULL;
+	int subnode = 0;
+
+	DMSG("Populating %s firewall bus", ctrl->name);
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		unsigned int i = 0;
+
+		if (fdt_get_status(fdt, subnode) == DT_STATUS_DISABLED)
+			continue;
+
+		if (IS_ENABLED(CFG_INSECURE) &&
+		    stm32mp_allow_probe_shared_device(fdt, subnode)) {
+			DMSG("Skipping firewall attributes check for %s",
+			     fdt_get_name(fdt, subnode, NULL));
+			goto skip_check;
+		}
+
+		DMSG("Acquiring firewall access for %s when probing bus",
+		     fdt_get_name(fdt, subnode, NULL));
+
+		do {
+			/*
+			 * The access-controllers property is mandatory for
+			 * firewall bus devices
+			 */
+			res = firewall_dt_get_by_index(fdt, subnode, i, &fw);
+			if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+				/* Stop when nothing more to parse */
+				break;
+			} else if (res) {
+				EMSG("%s: Error on node %s: %#"PRIx32,
+				     ctrl->name,
+				     fdt_get_name(fdt, subnode, NULL), res);
+				panic();
+			}
+
+			res = firewall_acquire_access(fw);
+			if (res) {
+				EMSG("%s: %s not accessible: %#"PRIx32,
+				     ctrl->name,
+				     fdt_get_name(fdt, subnode, NULL), res);
+				panic();
+			}
+
+			firewall_put(fw);
+			i++;
+		} while (true);
+
+skip_check:
+		res = dt_driver_maybe_add_probe_node(fdt, subnode);
+		if (res) {
+			EMSG("Failed on node %s with %#"PRIx32,
+			     fdt_get_name(fdt, subnode, NULL), res);
+			panic();
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result init_etzpc_from_dt(struct etzpc_device *etzpc_dev,
 				     const void *fdt, int node)
 {
@@ -689,7 +756,7 @@ static TEE_Result stm32_etzpc_probe(const void *fdt, int node,
 	if (res)
 		goto err;
 
-	res = firewall_dt_probe_bus(fdt, node, controller);
+	res = stm32_etzpc_dt_probe_bus(fdt, node, controller);
 	if (res)
 		goto err;
 
-- 
2.47.2

